src/mixins/createRenderMixin.ts

```diff
@@ -0,0 +1,289 @@
+	 */
+	render?: RenderFunction;
+
+	/**
+	 * Override the widget's tagName during construction
+	 */
+	tagName?: string;
+
+	/**
+	 * Set the widget's parent during construction
+	 */
+	parent?: Parent;
+}
+
+export interface RenderMixinFactory extends ComposeFactory<RenderMixin<RenderMixinState>, RenderMixinOptions<RenderMixinState>> {
+	idBase: string;
+}
+
+/**
+ * A map of dirty flags used when determining if the render function
+ * should be called
+ */
+const dirtyMap = new Map<RenderMixin<RenderMixinState>, boolean>();
+
+/**
+ * A weak map of the rendered VNode to return when the widget is
+ * not dirty.
+ */
+const renderCache = new WeakMap<RenderMixin<RenderMixinState>, VNode>();
+
+/**
+ * A weak map to shadow the classes for the widget
+ */
+const shadowClasses = new WeakMap<RenderMixin<RenderMixinState>, string[]>();
+
+/**
+ * A weak map to shadow the styles for the widget
+ */
+const shadowStyles = new WeakMap<RenderMixin<RenderMixinState>, StylesHash>();
+
+/**
+ * The counter for generating a unique ID
+ */
+let cachedRenderCount = 0;
+
+/**
+ * A function that generates an ID
+ */
+function generateID(cachedRender: RenderMixin<RenderMixinState>): string {
+	const id = `${createRenderMixin.idBase}${++cachedRenderCount}`;
+	cachedRender.setState({ id });
+	return id;
+}
+
+/**
+ * A weak map of the historic classes associated to a specific widget
+ */
+const widgetClassesMap = new WeakMap<RenderMixin<RenderMixinState>, string[]>();
+
+const createRenderMixin = createStateful
+	.mixin<Render, RenderMixinOptions<RenderMixinState>>({
+		mixin: {
+			get classes(this: RenderMixin<RenderMixinState>): string[] {
+				return (this.state && this.state.classes) || shadowClasses.get(this);
+			},
+
+			set classes(this: RenderMixin<RenderMixinState>, value: string[]) {
+				if (this.state.classes) {
+					this.setState({ classes: value });
+				}
+				else {
+					shadowClasses.set(this, value);
+					this.invalidate();
+				}
+			},
+
+			getNodeAttributes(this: RenderMixin<RenderMixinState>, overrides?: VNodeProperties): VNodeProperties {
+				const props: VNodeProperties = {};
+
+				this.nodeAttributes.forEach((fn) => {
+					const newProps: VNodeProperties = fn.call(this, assign({}, props));
+					if (newProps) {
+						assign(props, newProps);
+					}
+				});
+
+				if (overrides) {
+					assign(props, overrides);
+				}
+				return props;
+			},
+
+			getChildrenNodes(this: RenderMixin<RenderMixinState>): (VNode | string)[] {
+				return this.state.label ? [ this.state.label ] : undefined;
+			},
+
+			get id(this: RenderMixin<RenderMixinState>): string {
+				return (this.state && this.state.id) || generateID(this);
+			},
+
+			invalidate(this: RenderMixin<RenderMixinState>): void {
+				if (dirtyMap.get(this)) { /* short circuit if already dirty */
+					return;
+				}
+				const parent = this.parent;
+				dirtyMap.set(this, true);
+				renderCache.delete(this); /* Allow GC to occur on renderCache */
```

Comment:
In this situation, the instance will persist and therefore the cache will persist until the cache is set again (which may never occur)... Therefore an invalidated cache would not be freed up for GC until a) the instance is dereferenced or b) the widget is rendered again.  Assuming you had a large part of an application that is not visible but has been invalidated, that could be quite a large cache which you could free up for GC.